src/diff/oasDiff.ts

```diff
@@ -4,8 +4,183 @@
+import path from "path";
+
+/**
+ * Interface for flags used in oasDiff functions
+ */
+interface OasDiffFlags {
+  format?: "json" | "console";
+  dir?: boolean;
+  "out-file"?: string;
+}
+
+/**
+ * Represents a directory-level API change (added/deleted)
+ */
+interface DirectoryStatusChange {
+  directory: string;
+  status: "added" | "deleted";
+  message: string;
+}
+
+/**
+ * Represents file-level changes within a directory
+ */
+interface FileChange {
+  file: string;
+  status: "added" | "deleted";
+  message: string;
+}
+
+/**
+ * Represents changes found in a directory with detailed change information
+ */
+interface DirectoryChanges {
+  directory: string;
+  changes: (FileChange | Record<string, unknown>)[];
+}
+
+/**
+ * Union type for all possible result items
+ * - string: Text format output
+ * - DirectoryStatusChange: JSON format for added/deleted directories
+ * - DirectoryChanges: JSON format for directories with file changes
+ * - unknown[]: Parsed JSON output from oasdiff (structure depends on oasdiff output)
+ */
+type OasDiffResult =
+  | string
+  | DirectoryStatusChange
+  | DirectoryChanges
+  | unknown[];
+
+/**
+ * Return type for oasDiff handler functions
+ */
+interface OasDiffHandlerResult {
+  results: OasDiffResult[];
+  hasChanges: boolean;
+  hasErrors: boolean;
+}
+
+/**
+ * Recursively find directories containing exchange.json files
+ *
+ * @param rootPath - The root path to search from
+ * @returns Array of objects with directory name and full path
+ */
+async function findExchangeDirectories(
+  rootPath: string
+): Promise<Array<{ name: string; path: string }>> {
+  const result = [];
+  // assume the directory depth is 3, we don't want to go too deep
+  const maxDepth = 3;
+  let foundAnyExchangeJson = false;
+
+  async function searchDirectory(currentPath: string, depth = 0) {
+    if (depth > maxDepth) {
+      throw new Error(
+        `Maximum directory depth (${maxDepth}) exceeded while searching for exchange.json files in: ${currentPath}`
+      );
+    }
+
+    try {
+      const entries = await fs.readdir(currentPath);
+
+      // Check if current directory contains exchange.json
+      const hasExchangeJson = entries.includes("exchange.json");
+
+      if (hasExchangeJson) {
+        foundAnyExchangeJson = true;
+        const dirName = path.basename(currentPath);
+        result.push({
+          name: dirName,
+          path: currentPath,
+        });
+        return;
+      }
+
+      // Check if this is a leaf directory (no subdirectories)
+      const subdirectories = [];
+      for (const entry of entries) {
+        const entryPath = path.join(currentPath, entry);
+        const stat = await fs.stat(entryPath);
+        if (stat.isDirectory()) {
+          subdirectories.push(entryPath);
+        }
+      }
+      // If this is a leaf directory and we haven't found exchange.json, that's an error
+      if (subdirectories.length === 0 && !hasExchangeJson) {
+        throw new Error(
```

Comment:
I think we can log it as a warning and proceed instead of throwing an error and exiting